metricbeat/module/linux/ksm/ksm.go

```diff
@@ -0,0 +1,81 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package ksm
+
+import (
+	"path/filepath"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/metricbeat/module/linux"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("linux", "ksm", New)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	fs string
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The linux pageinfo metricset is beta.")
+	linuxModule, ok := base.Module().(*linux.Module)
+	if !ok {
+		return nil, errors.New("unexpected module type")
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		fs:            filepath.Join(linuxModule.HostFS, "/sys/kernel/mm/ksm"),
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+
+	ksmData, err := fetchKSMStats(m.fs)
+	if err != nil {
+		return errors.Wrap(err, "error fetching KSM stats")
+	}
+
+	report.Event(mb.Event{
+		MetricSetFields: common.MapStr{
+			"stats": ksmData,
```

Comment:
`ksmData` has type `ksmData`. libbeat unfortunately require `common.MapStr`, such that events can be used with processors. Is there some conversation happening somewhere?

Reply:
Not sure what you mean? I could swear I've seen struct map values in other parts of metricbeat, but maybe I'm wrong?

Reply:
Oof... TIL

Libbeat tries to normalize events on intake, but only understands some 'normalized' types like `int,` `bool`, `string`, `map[string]interface{}`, `time.Time`...

If it finds a type it does not understand, it json encodes the event into a temporary buffer, and decodes it into `map[string]interface{}`. Yeah, structs with json struct tags actually work :/